### PR TITLE
Balance: Invert Scale and Buff Quad Cannon Guns DPS

### PR DESCRIPTION
**Pull request by xezon**
Reference issue:
* #56

## Original
before

## Patched
quad_3_bonus

## Original
https://user-images.githubusercontent.com/4720891/182035421-266e6e12-fce2-44ca-a4f5-2abe67ba91e7.mp4

## Patched
https://user-images.githubusercontent.com/4720891/182035466-b0b29c02-b4e4-4677-bc95-212a2c8a151e.mp4

### CODE REVIEW

**Comment by xezon** [comment]
@Jundiyy Can you upload you Quad Damage test map please?

File: Patch104pZH/GameFilesEdited/Data/INI/Weapon.ini [hunk 1]
```diff
@@ -2699,6 +2708,15 @@ Weapon QuadCannonGunUpgradeOne
   AntiAirborneVehicle   = No
   AntiAirborneInfantry  = No
   WeaponBonus           = PLAYER_UPGRADE DAMAGE 125% ; APBullets
+  WeaponBonus           = HORDE              DAMAGE 112.5% ; Patch104p @balance Total DPS Bonus  60% to  80%
```
Comment (commy2) [inline Patch104pZH/GameFilesEdited/Data/INI/Weapon.ini]: HORDE ? Quad cannot horde.

**Comment by commy2** [comment]
How did you arrive at these values for NATIONALISM, FANATICISM, ENTHUSIASTIC (China) and TARGET_FAERIE_FIRE (USA)?

File: Patch104pZH/GameFilesEdited/Data/INI/Weapon.ini (hunk 1 above)
Comment (xezon) [reply Patch104pZH/GameFilesEdited/Data/INI/Weapon.ini]: Yeah is not used, but does not hurt to have it here, just in case someone decides to create a Mod based on our Patch to add Quad Hordes :-)

**Comment by xezon** [comment]
In first table above, you see see damage 100 60 60, the damage % tweak redistributes it to 60 80 80. The calculation are done via weapon_dps_calculator in https://github.com/TheSuperHackers/GeneralsGamePatch/tree/main/Patch104pZH/Design/Balancing

**Comment by commy2** [comment]
Wouldn't it be easier to lock the Quad at one shot per 2 frames (unscrapped) and one shot per 1 frame (scrapped) and then give reasonable damage buffs?

**Comment by xezon** [comment]
From what I gathered weapon tick is fixed at every 2 frames, respectively 1/15 of a second or 66.66 ms. It cannot go lower. Quad already shoots at this lowest rate with the unscrapped weapon.

TLDR
We can't do that.

**Comment by commy2** [comment]
Unscrapped without vet actually shoots every 2 frames. Any ROF bonus and it goes to every 1 frame.

My idea would be to invert all ROF bonuses on the unscrapped quad, and then compensate with reasonable damage buffs.

**Comment by xezon** [comment]
Unscrapped without vet shoots every 4 frames. With ROF bonus it goes to 2 frames. I am all ears for new suggestions, because this proposal is still not ideal, when certain bonuses stack. For example check HERO+BATTLEPLAN_BOMBARDMENT, QuadCannonGun deals more damage than QuadCannonGunUpgradeTwo.

**Comment by alanblack166** [comment]
Better to just make the quad cannon fire every 2 frames and rescale the damage accordingly. Veterancy can be compensated with extra damage per shot.

**Comment by commy2** [comment]
> Unscrapped without vet shoots every 4 frames. With ROF bonus it goes to 2 frames

I was talking about simulation frames. There are twice as many draw frames. (30 vs 15 per second) So it was a misunderstanding.

**Comment by xezon** [comment]
> Better to just make the quad cannon fire every 2 frames and rescale the damage accordingly. Veterancy can be compensated with extra damage per shot.

It is not clear how that would work. We need very accurate values for this. Reducing ROF changes a lot of parameters.

**Comment by xezon** [comment]
I looked at the Excel sheet for another hour and tried tweaking some values, but its madness. Changing as little as possible of the base values is the most sane approach I think. Plus the differences on some stacked bonuses are not that big, so overall this change looks acceptable to me.

Here is diff from the new sheet.

(Redacted sheets. Contained Errors.)

**Comment by commy2** [comment]
Why don't you start with the RATE_OF_FIRE changes from this PR:

https://github.com/TheSuperHackers/GeneralsGamePatch/pull/359/files

This locks the Quad to 1 shot per 2 logic frames without scrap and to 1 shot per 1 logic frame with one or two scraps, regradless of any ROF upgrade.

You can then give it a DAMAGE buff to whatever DPS increase you want, without any math required at all.

**Comment by xezon** [comment]
I do not understand why you suggest that. To follow your inquiry, I punched the numbers of your "Generic Vet" into the DPS calculator, and this is the result:

(Redacted. Wrong data.)

~~This is not good distribution. Touching Fire Rate makes it an absolute mess.~~

~~Also: Changing fire rate will also affect the sound playback interval. And the bullet tracers. And muzzle flashes. I do not think Fire Rate should be touched at all here. At least not for the values as this change intended to tweak to (which it does).~~

**Comment by xezon** [comment]
Additionally, I think this change is better on paper:
* #728

**Comment by xezon** [comment]
I fixed the logic issue in the DPS calculator, and recalculated the correct modifiers for the intended target values. New comparison videos have been added to first post.

**Comment by xezon** [comment]
@commy2 
If you want to change ROF, I suggest to create new Pull Request. I don't know how to do it gracefully. That said, I am happy with the numbers here as is now, albeit not perfect.
